lib/worker_test.go

```diff
@@ -0,0 +1,235 @@
+package lib
+
+import (
+	"context"
+	"encoding/json"
+	"testing"
+	"time"
+
+	"github.com/hanfei1991/microcosm/pkg/adapter"
+	"github.com/hanfei1991/microcosm/pkg/clock"
+	"github.com/hanfei1991/microcosm/pkg/metadata"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+func putMasterMeta(ctx context.Context, t *testing.T, metaclient metadata.MetaKV, metaData *MasterMetaKVData) {
+	masterKey := adapter.MasterInfoKey.Encode(masterName)
+	masterInfoBytes, err := json.Marshal(metaData)
+	require.NoError(t, err)
+	_, err = metaclient.Put(ctx, masterKey, string(masterInfoBytes))
+	require.NoError(t, err)
+}
+
+func TestWorkerInitAndClose(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	worker := newMockWorkerImpl(workerID1, masterName)
+	worker.clock = clock.NewMock()
+	worker.clock.(*clock.Mock).Set(time.Now())
+	putMasterMeta(ctx, t, worker.metaKVClient, &MasterMetaKVData{
+		ID:          masterName,
+		NodeID:      masterNodeName,
+		Epoch:       1,
+		Initialized: true,
+	})
+
+	worker.On("InitImpl", mock.Anything).Return(nil)
+	worker.On("Status").Return(WorkerStatus{
+		Code: WorkerStatusNormal,
+	}, nil)
+	err := worker.Init(ctx)
+	require.NoError(t, err)
+
+	worker.clock.(*clock.Mock).Add(defaultTimeoutConfig.workerHeartbeatInterval + 1*time.Second)
+	worker.clock.(*clock.Mock).Add(defaultTimeoutConfig.workerHeartbeatInterval + 1*time.Second)
```

Comment:
These two lines are the same, is it by design?

Reply:
Yes.

Reply:
It seems that the mock clock's ticker needs this to trigger a tick.